File: Haskell/N-Queens.hs
-- Resenje problema N kraljica.
-- Tabla je matrica sa 0 i 1

validna::[[Int]] -> Int -> Bool
validna tabla n = and [validanRed tabla n 0 i | i <- [1..(n-1) ]]

validanRed:: [[Int]] -> Int -> Int -> Int -> Bool 
validanRed tabla n prethodni i 
    | prethodni == i = True
    | (mestoP) == (mesto) = False 
    | (abs (mestoP - mesto)) == (abs (i - prethodni)) = False
    | otherwise = validanRed tabla n (prethodni+1) i
    where
        mesto = (snd . head) ( filter (\x -> ((fst x) == 1) ) (zip (tabla !! i) [0 .. (n-1)] )) 
        mestoP = (snd . head) ( filter (\x -> ((fst x) == 1) ) (zip (tabla !! prethodni) [0 .. (n-1)]) )


generisiSve:: Int -> Int -> [[[Int]]]
generisiSve n red
    | red == (n) = [[]] 
    | otherwise = resenje
    where
        mogucnosti = [((take i [0, 0..]) ++ (take (n-i) (1:[0, 0..])))| i <- [0..n-1]] 
        ostatak = (generisiSve n (red + 1))
        resenje = [ [prvi] ++ tabla |prvi <- mogucnosti, tabla <- ostatak]

kraljice:: Int -> [[[Int]]]
kraljice n = filter (\x -> (validna x n)) (generisiSve n 0)

File: Scala/Nqueens.scala
/*
    Resenje problema N kraljica.
*/

object Main {

    
    def jeValidno(konf:List[Int], nivo:Int,pot:Int):Boolean = {
        if(nivo == 0) {
            return true
        }

        val doSada= konf.take(nivo)

        doSada.zip((0 until nivo).toList).forall(x => (x._1) != pot && (Math.abs(x._1 - pot) != Math.abs(x._2 - nivo)))
    }

    def nqueens(trenutno:List[Int], nivo:Int, n:Int):List[List[Int]] = {
        if(nivo == n) {
            return List(trenutno)
        }

        var res:List[List[Int]] = Nil
        for(indeks <- (0 until n).toList if jeValidno(trenutno,nivo,indeks)) {  
            res = res:::nqueens(trenutno:::List(indeks), nivo + 1, n)
        }
        res
        
    }
    def nqueens(n:Int):List[List[Int]] = {
       nqueens(Nil, 0, n)
    }

    def main(args: Array[String]): Unit = {

        println(nqueens(4))
    }
}